version/version.go

```diff
@@ -1,57 +1,127 @@
+	// for each release. The main branch is always the next minor version
+	// compared to the latest release branch.
+	Major = 2
+	Minor = 1
+
+	// The patch version, which must only be non-zero for release candidates
+	// and official releases. All development versions must have a patch
+	// version of 0.
+	Patch = 0
+
+	// The pre-release version. This is set to "dev" for development versions
+	// on the main branch and should be updated to "rcX" for release candidates
+	// on release branches only. For a final release, this must be set to an
+	// empty string. All other values are invalid.
+	PreRelease = "dev"
 )
 
-// Version holds the textual version string.
-var Version = func() string {
-	return fmt.Sprintf("%d.%d.%d", VersionMajor, VersionMinor, VersionPatch)
-}()
+// Get returns the complete version information.
+func Get() Version {
+	return _version
+}
 
-// VersionWithMeta holds the textual version string including the metadata.
-var VersionWithMeta = func() string {
-	v := Version
-	if VersionMeta != "" {
-		v += "-" + VersionMeta
-	}
-	return v
-}()
+// String returns the version string.
+func String() string {
+	return Get().String()
+}
 
-func VersionWithCommit(gitCommit, gitDate string) string {
-	vsn := VersionWithMeta
+// StringWithCommit returns the version string with the commit hash and date.
+func StringWithCommit() string {
+	vsn := Get().String()
 	if len(gitCommit) >= 8 {
 		vsn += "-" + gitCommit[:8]
 	}
-	if (VersionMeta != "stable") && (gitDate != "") {
+	if gitDate != "" {
 		vsn += "-" + gitDate
 	}
 	return vsn
 }
 
-func AsString() string {
-	return ToString(uint16(VersionMajor), uint16(VersionMinor), uint16(VersionPatch))
+// GitCommit returns the commit hash if available. If not, the empty string
+// is returned.
+func GitCommit() string {
+	return gitCommit
 }
 
-func AsU64() uint64 {
-	return ToU64(uint16(VersionMajor), uint16(VersionMinor), uint16(VersionPatch))
+// GitDate returns the commit date if available. If not, the empty string
+// is returned.
+func GitDate() string {
+	return gitDate
 }
 
-func ToU64(vMajor, vMinor, vPatch uint16) uint64 {
-	return uint64(vMajor)*1e12 + uint64(vMinor)*1e6 + uint64(vPatch)
+// Version represents a version of the code.
+type Version struct {
+	Major      int
+	Minor      int
+	Patch      int
+	PreRelease string
 }
 
-func ToString(major, minor, patch uint16) string {
-	return fmt.Sprintf("%d.%d.%d", major, minor, patch)
+// IsRelease returns true if the version is a release version. It returns false
+// if the version has a meta string or is dirty.
+func (v Version) IsRelease() bool {
+	return v.PreRelease == ""
 }
 
-func U64ToString(v uint64) string {
-	return ToString(uint16((v/1e12)%1e6), uint16((v/1e6)%1e6), uint16(v%1e6))
+func (v Version) String() string {
+	res := fmt.Sprintf("%d.%d.%d", v.Major, v.Minor, v.Patch)
+	if v.PreRelease != "" {
+		res += "-" + v.PreRelease
+	}
+	return res
 }
+
+var preReleaseRE = regexp.MustCompile(`^(|dev|rc(\d+))$`)
```

Comment:
nit: declare all global variables together?

Reply:
Since this is just a performance tweak for the function following immediately below, and not a global variable supposed to be used by anybody else, I think it should be defined just next to it. The section of constants defined on the top, to the contrary, are supposed to be public constants for users of the package. No need to mix them up.